core/frontend/src/components/autopilot/FirmwareManager.vue

```diff
@@ -0,0 +1,269 @@
+import { getModule } from 'vuex-module-decorators'
+
+import AutopilotStore from '@/store/autopilot'
+import NotificationStore from '@/store/notifications'
+import { Firmware, Vehicle } from '@/types/autopilot'
+import { autopilot_service } from '@/types/frontend_services'
+import { LiveNotification, NotificationLevel } from '@/types/notifications'
+
+const notification_store: NotificationStore = getModule(NotificationStore)
+const autopilot_store: AutopilotStore = getModule(AutopilotStore)
+
+enum UploadType {
+  Cloud,
+  File
+}
+
+export default Vue.extend({
+  name: 'FirmwareManager',
+  data() {
+    return {
+      UploadType,
+      upload_type: UploadType.Cloud,
+      chosen_vehicle: null as Vehicle | null,
+      chosen_firmware_url: null as URL | null,
+      available_firmwares: [] as Firmware[],
+      cloud_firmware_available: false,
+      file_firmware_available: false,
+      firmware_file: null as Blob | null,
+      installing: false,
+      install_completed: false,
+      successfull_install: false,
+      install_result_message: '',
+    }
+  },
+  computed: {
+    install_alert_color(): string {
+      return this.successfull_install ? 'teal' : 'red'
+    },
+    install_alert_icon(): string {
+      return this.successfull_install ? 'mdi-check-decagram' : 'close-octagon'
+    },
+    vehicle_types(): {value: string, text: Vehicle}[] {
+      return Object.entries(Vehicle).map(
+        (vehicle) => ({ value: vehicle[0], text: vehicle[1] }),
+      )
+    },
+    disable_firmware_selection(): boolean {
+      return this.chosen_vehicle == null
+    },
+    showable_firmwares(): {value: URL, text: string}[] {
+      return this.available_firmwares.map(
+        (firmware) => ({ value: firmware.url, text: firmware.name }),
+      )
+    },
+    allow_installing(): boolean {
+      if (this.installing) {
+        return false
+      }
+      return this.upload_type === UploadType.Cloud ? this.cloud_firmware_available : this.file_firmware_available
+    },
+  },
+  methods: {
+    async updateAvailableFirmwares(): Promise<void> {
+      await axios({
+        method: 'get',
+        url: `${autopilot_store.API_URL}/available_firmwares`,
+        timeout: 2000,
+        params: { vehicle: this.chosen_vehicle },
+      })
+        .then((response) => {
+          this.available_firmwares = response.data
+        })
+        .catch((error) => {
+          notification_store.pushNotification(new LiveNotification(
+            NotificationLevel.Error,
+            autopilot_service,
+            'FIRMWARE_FETCH_FAIL',
+            `Could not fetch available firmwares: ${error.message}.`,
+          ))
+        })
+    },
+    setCloudFirmwareAvailable(): void {
+      this.cloud_firmware_available = true
+    },
+    setFileFirmwareAvailable(file: Blob): void {
+      this.firmware_file = file
+      this.file_firmware_available = true
+    },
+    async installFirmware(): Promise<void> {
+      this.installing = true
```

Comment:
This looks like an enum disguised as a series of booleans

Reply:
Definitely. Will make an Enum for install status.